pymanopt/tools/diagnostics.py

```diff
@@ -0,0 +1,201 @@
+except ImportError:
+    plt = None
+
+
+def identify_linear_piece(x, y, window_length):
+    """Identify a segment of the curve (x, y) that appears to be linear.
+    This function attempts to identify a contiguous segment of the curve
+    defined by the vectors x and y that appears to be linear. A line is fit
+    through the data over all windows of length window_length and the best
+    fit is retained. The output specifies the range of indices such that
+    x(segment) is the portion over which (x, y) is the most linear and the
+    output poly specifies a first order polynomial that best fits (x, y) over
+    that segment, following the usual matlab convention for polynomials
+    (highest degree coefficients first).
+    See also: check_directional_derivative check_gradient check_hessian
+    """
+    residues = np.zeros(len(x)-window_length)
+    polys = np.zeros(shape=(2, len(residues)))
+    for i in range(len(residues)):
+        segment = range(i, (i+window_length)+1)
+        poly, residuals, _, _, _ = np.polyfit(x[segment], y[segment],
+                                              1, full=True)
+        residues[i] = la.norm(residuals)
+        polys[:, i] = poly
+    best = np.argmin(residues)
+    segment = range(best, best+window_length+1)
+    poly = polys[:, best]
+    return segment, poly
+
+
+def get_directional_derivative(problem, x, d):
+    """Computes the directional derivative of the cost function at x along d.
+    Returns the derivative at x along d of the cost function described in the
+    problem structure.
+    """
+    if hasattr(problem.manifold, "diff"):
+        diff = problem.manifold.diff(x, d)
+    else:
+        grad = problem.manifold.grad(x)
+        diff = problem.manifold.inner(x, grad, d)
+    return diff
+
+
+def check_directional_derivative(problem, x=None, d=None,
+                                 force_gradient=False):
+    """Checks the consistency of the cost function and directional derivatives.
+    check_directional_derivative performs a numerical test to check that the
+    directional derivatives defined in the problem structure agree up to first
+    order with the cost function at some point x, along some direction d. The
+    test is based on a truncated Taylor series (see online pymanopt
+    documentation).
+    Both x and d are optional and will be sampled at random if omitted.
+    See also: check_gradient check_hessian
+    If force_gradient is True, then the function will call get_gradient and
+    infer the directional derivative, rather than call
+    get_directional_derivative directly. This is used by check_gradient.
+    """
+    #  If x and / or d are not specified, pick them at random.
+    if d is not None and x is None:
+        raise ValueError("If d is provided, x must be too, "
+                         "since d is tangent at x.")
+    if x is None:
+        x = problem.manifold.rand()
+    elif x.shape != problem.manifold.rand().shape:
```

Comment:
The `rand()` method doesn't always return a numpy array. For the `FixedRankEmbedded` manifold for instance it returns a tuple which doesn't have a shape attribute. I'd prefer to remove these shape checks and subsequent reshape attempts altogether. If a user passed in bogus points, it's okay to let the first shape-mismatch exceptions bubble up wherever it may trigger first.